### DISCUSSION TITLE
Intergrated Terminal

### DISCUSSION THREAD

**Discussion by mixstam1821** [orbitalquark/textadept#497] [Ideas]
Hi, textadept is a fantastic editor. I wonder if it would be possible to have an integrated terminal (as vscode and other editors have) where I could send my selected code to run, etc. I found how to send it in tmux (https://github.com/orbitalquark/textadept/wiki/selection2tmux), but I would like a more cross-platform solution, for example, to send code in konsole (linux) or powershell (windows).

**Comment by orbitalquark** [comment]
It's possible, but would be very difficult.

I've tried to do this a few times using a buffer via a Lua library that 
creates a pseudo-terminal (https://github.com/AshleighAdams/lua-pty), 
but it's a LOT of work to turn it into a usable emulator. Each time I 
pick it back up, I run into a new difficulty, most recently dealing with 
terminal escape sequences to try and show colors.

And that's just Linux. I'm not sure how I'd manage to do this on Windows :(

**Reply by mixstam1821** [reply]
Hi, thanks for your response.

Finally, with this in my init.lua I can send a selection or the current line (and the cursor moves down) to an active Konsole (Linux) terminal (not inside the textadept):

``
local function send_and_execute_in_konsole()
    local start_pos = buffer.current_pos
    local end_pos = buffer.current_pos
    local txt = buffer:get_sel_text()

    -- Check if text is not empty
    if txt and #txt > 0 then
        -- Remove all consecutive empty lines
        txt = txt:gsub('\n+', '\n')

        -- If the last line is empty, remove it
        txt = txt:gsub('\n$', '')

        -- Quote the text to handle special characters
        txt = "'" .. txt:gsub("'", "'\\''") .. '\n' .. "'"
    else
        -- If no text is selected, execute the current line
        local line_num = buffer:line_from_position(buffer.current_pos)
        txt = "'" .. buffer:get_line(line_num):gsub("'", "'\\''") .. '\n' .. "'"
    end

    -- Use qdbus to send and execute text in Konsole
    local command = "qdbus org.kde.konsole-$(pidof konsole) /Sessions/1 sendText " .. txt
    local handle = io.popen(command)
    handle:close()

    -- Move the cursor down
    buffer:set_selection(end_pos, start_pos)
    buffer:line_down()
end

-- Assign a keybinding (ctrl+.) to the send_and_execute_in_konsole function
keys['ctrl+.'] = send_and_execute_in_konsole
``

With this function also, I can send to ipython, a function definition (def main(): ....) without breaking down my function's code if it contains empty lines.

I hope, it helps.
Cheers.

**Reply by orbitalquark** [reply]
Cool! Thanks for sharing :)

**Comment by alurm** [comment]
Having a bare bones terminal emulator would be useful too. Acme text editor comes with one, I use it all the time (well, I use Plan 9 port's Acme, technically).

https://man.cat-v.org/plan_9/1/acme, search for "win ".

So, on Unix, if you set TERM to dumb, PAGER to cat, and NO_COLOR to 1, you can get quite far.

Here is win.c, it's only nine hundred lines of code: https://github.com/9fans/plan9port/blob/master/src/cmd/9term/win.c.

**Comment by dcst-it** [comment]
900 lines of code for the integrated terminal emulator "only" is almost the same number of lines of the entire C source code of Textadept :) at least judging from the code stats graph in the home page.

I'm another one who uses a lot the system terminal emulator, as I run `make` commands all the time. I found a convenient compromise by implementing in my `init.lua` configuration file a function based on the following solution present in the Textadept wiki: <https://github.com/orbitalquark/textadept/wiki/TerminalHere>. The function is bound to the `Control + t` shortcut: a simple keypress and the external terminal emulator magically opens in the directory of the current file. Super useful. I also bound another shortcut to a similar function which launches the system file manager in the directory of the current file.

This is enough for me, as I can use Textadept for what it does best (text exiting) and a dedicated terminal emulator with all of its bells and whistles (such as multiple tabs and glorious COLORS :)) for shell commands. Moreover, having separate windows for the editor and the terminal sometimes is nice - and of course sometimes is annoying, but:

- with large screens two windows can be visualized at the same time, while with small laptop screens integrated terminals and other GUI panels tend to steal space from the editor view;
- with the OS keyboard shortcuts is always possible to switch between the two windows without using the mouse.

So while having an integrated terminal would surely be nice, the implementation and maintenance burden would probably overweight the actual value of the feature...

As for Windows, for example the Geany editor/IDE offers an integrated terminal emulator on Linux, but the feature is not present in the Windows version.

**Comment by Fwirt** [comment]
Another good example of a "minimal" terminal codebase is [st](https://st.suckless.org/) but terminal emulators are generally insane pieces of software that are just emulating the functionality of antique hardware.

A better approach would probably be something akin to [Eshell](https://www.gnu.org/software/emacs/manual/html_node/eshell/index.html#Top), which implements a shell like bash rather than a terminal emulator. The added advantage of this is that you can call Textadept functions from the "shell" and pipe the output of commands back into Textadept functions. The editor behaves more like a line printer than a terminal emulator, so you can't use curses apps and stuff like that, but that seems better suited to dedicated terminal emulator software. Using a Scintilla as a dumb character display would sort of remove a lot of incentive for using Scintilla in the first place.
